libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,727 @@
+*    by two pairs of sequential points, and then cuts out anything between two
+*    points when their line segments get close. This algorithm will never
+*    compare two consecutive line segments. Obviously the segments (p1,p2) and
+*    (p2,p3) will get very close (they touch), but there would be nothing to
+*    trim between them.
+*
+*    2. Simplification uses the Ramer-Douglas-Peucker algorithm. See Wikipedia
+*    for a more complete description.
+*
+*    The simplification and pruning algorithms run in the background and do not
+*    alter the path in memory.  Two definitions, SAFERTL_SIMPLIFY_TIME_US and
+*    SAFERTL_PRUNING_LOOP_TIME_US are used to limit how long each algorithm will
+*    be run before they save their state and return.
+*
+*    Both algorithms are "anytime algorithms" meaning they can be interrupted
+*    before they complete which is helpful when memory is filling up and we just
+*    need to quickly identify a handful of points which can be deleted.
+*
+*    Once the algorithms have completed the simplify.complete and
+*    prune.complete flags are set to true.  The "thorough cleanup" procedure,
+*    which is run as the vehicle initiates the SafeRTL flight mode, waits for
+*    these flags to become true.  This can force the vehicle to pause for a few
+*    seconds before initiating the return journey.
+*/
+
+AP_SafeRTL::AP_SafeRTL(const AP_AHRS& ahrs, bool example_mode) :
+    _ahrs(ahrs),
+    _example_mode(example_mode)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplify.bitmask.setall();
+}
+
+// initialise safe rtl including setting up background processes
+void AP_SafeRTL::init()
+{
+    // protect against repeated call to init
+    if (_path != nullptr) {
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _points_max = constrain_int16(_points_max, 0, SAFERTL_POINTS_MAX);
+
+    // check if user has disabled SafeRTL
+    if (_points_max == 0 || !is_positive(_accuracy)) {
+        return;
+    }
+
+    // allocate arrays
+    _path = (Vector3f*)calloc(_points_max, sizeof(Vector3f));
+
+    _prune.loops_max = _points_max * SAFERTL_PRUNING_LOOP_BUFFER_LEN_MULT;
+    _prune.loops = (prune_loop_t*)calloc(_prune.loops_max, sizeof(prune_loop_t));
+
+    _simplify.stack_max = _points_max * SAFERTL_SIMPLIFY_STACK_LEN_MULT;
+    _simplify.stack = (simplify_start_finish_t*)calloc(_simplify.stack_max, sizeof(simplify_start_finish_t));
+
+    // check if memory allocation failed
+    if (_path == nullptr || _prune.loops == nullptr || _simplify.stack == nullptr) {
+        log_action(SRTL_DEACTIVATED_INIT_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING, "SafeRTL deactivated: init failed");
+        free(_path);
+        free(_prune.loops);
+        free(_simplify.stack);
+        return;
+    }
+
+    _path_points_max = _points_max;
+
+    // create semaphore
+    _path_sem = hal.util->new_semaphore();
```

Comment:
Technically can be null if we perfectly used up memory, might as well do this before checking all the other nullptr's